main.cpp

```diff
@@ -179,147 +134,532 @@ int main()
+DoubleType::DoubleType(double input)
 {
-    return lhs + rhs;
+    value = new double(input);
 }
 
-double DoubleType::subtract(double lhs, double rhs)
+DoubleType::~DoubleType()
 {
-    return lhs - rhs;
+    delete value;
 }
 
-double DoubleType::multiply(double lhs, double rhs)
+DoubleType& DoubleType::add(double input)
 {
-    return lhs * rhs;
+    *value += input;
+    return *this;
 }
 
-double DoubleType::divide(double lhs, double rhs)
+DoubleType& DoubleType::add(const FloatType& input)
 {
-    if(rhs == 0.0) 
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !" << std::endl;
+    *value += *input.value;
+    return *this;
+}
 
-    return lhs / rhs;
+DoubleType& DoubleType::add(const DoubleType& input)
+{
+    *value += *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::add(const IntType& input)
+{
+    *value += *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(double input)
+{
+    *value -= input;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const FloatType& input)
+{
+    *value -= *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const DoubleType& input)
+{
+    *value -= *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const IntType& input)
+{
+    *value -= *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(double input)
+{
+    *value *= input;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const FloatType& input)
+{
+    *value *= *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const DoubleType& input)
+{
+    *value *= *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const IntType& input)
+{
+    *value *= *input.value;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double input)
+{
+    if(input == 0.0) 
+        std::cout << "warning: floating point division by zero!" << std::endl;
+
+    *value /= input;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const FloatType& input)
+{
+    float ftValue = *input.value;
+
+    if(ftValue == 0.f) 
+        std::cout << "warning: floating point division by zero!" << std::endl;
+
+    *value /= ftValue;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const DoubleType& input)
+{
+    double dtValue = *input.value;
+
+    if(dtValue == 0.0) 
+        std::cout << "warning: floating point division by zero!" << std::endl;
+
+    *value /= dtValue;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const IntType& input)
+{
+    int itValue = *input.value;
+
+    if(itValue == 0) 
+    {
+        std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+        
+        return *this;
+    }
+
+    *value /= itValue;
+    return *this;
 }
 
 // =======INTTYPE FUNCTIONS DEFINITIONS====================
 
-int IntType::add(int lhs, int rhs)
+IntType::IntType(int input)
+{
+    value = new int(input);
+}
```

Comment:
When you have to choose where to initialize your member variables and your choices are:
- Constructor Member Initializer List
- Inside the Constructor's body

Try to use the Constructor Member Initializer List.
The reason is that the member variables need to be instantiated/initialized when the class is created.
When you assign to them in the constructor body, you are potentially re-initializing them, which is a waste of resources.
Take a look at the output of this program: https://godbolt.org/z/q83YM63ze
Take note of the second instance being created and destroyed due to that constructor body assignment.
- [ ] revise, use the member initializer list to initialize your member variables